Stabilize migration from Metaschema repository (#3)
usnistgov/metaschema-xslt: README.md

# Metaschema XSLT

## License

See the [project license](LICENSE.md) in this repository.

This project is placed into the worldwide public domain.

## Software description

Metaschema-XSLT is an XSLT-based implementation of the NIST (ITL/CSD) [Metaschema Information Modeling Framework](https://github.com/usnistgov/metaschema).

NIST Metaschema defines how to describe sets of data objects -- serialized information sets, considered abstractly -- *across* and *between* implementations of processing systems that are coded to manipulate and process such objects.

> These descriptions are provided by means of a formal *language* developed for this purpose. In support of metaschemas defined using this language, this software can be used to generate utilities and software artifacts including XML Schema (XSD) and JSON Schema, data conversion scripts, and model documentation.

Reviewer: Editorial:
```suggestion
NIST Metaschema defines how to describe sets of data objects -- serialized information sets, considered abstractly -- *across* and *between* implementations of processing systems that are coded to manipulate and process such objects. These descriptions are provided by means of a formal *language* developed for this purpose.

In support of metaschemas defined using this language, this software can be used to generate utilities and software artifacts including XML Schema (XSD) and JSON Schema, data conversion scripts, and model documentation.
```

Author: Fixed in incoming commit.

The current effort seeks to implement, demonstrate and enable the capabilities of a NIST Metaschema system, with respect to functionalities including schema generation, data conversion or pipelining, stylesheet generation or a host of other potential uses of Metaschema and metaschemas considered broadly. In addition, we aspire to do so sustainably, not in the sense that the software will meet all requirements for all time, but that it will be reliable, validable, traceable, extensible -- and replaceable.

Considering data traceability and process transparency as primary goals will also serve secondary goals, such as the creation and demonstration of use cases (for this and compatible software) and the codification and maturation of relevant standards, including standards supported by Metaschema such as OSCAL.

Stress is therefore on:

- Transparency, testability and maintainability of processing on an open-source stack
- Conformance in support of Metaschema-based data interchange / interop
- Generality of support of Metaschema features

> What we do not emphasize:

Author: ```Suggestion
Additionally, we care about, but do not prioritize:

- Performance
- Portability beyond the XML/XSLT dependency stack (instead relying on other Metaschema initiatives)
- Specific use cases
```

- Performance
- Portability beyond the XML/XSLT dependency stack (instead relying on other Metaschema initiatives)

> - Use cases in isolation

Reviewer: Not sure what this means?

Author: Fixed in incoming commit.

### Origins

Formerly housed in the Metaschema repository, this code base traces the history of development of the Metaschema concept in the context of the OSCAL project. It was originally conceived as a demonstration and proof of concept, providing a bridge enabling JSON- and XML-based development in parallel over common problem sets and common data. Success in this effort led to a determination that multiple implementations of a platform-independent specification were needed, at which point this implementation was carved out into its own repository.

### Project sunset

> The NIST team has no long-term plan to maintain this project outside the context of the initiatives it supports, OSCAL[https://pages.nist.gov/OSCAL](https://pages.nist.gov/OSCAL) and Metaschema[https://pages.nist.gov/metaschema](https://pages.nist.gov/metaschema), while we reserve the right to change and adapt to community needs in alignment with NIST's mission. A project goal is to enable the stabilization and socialization of the Metaschema technology, as evidenced by the use and support of *other* Metaschema implementations.

Author: ```Suggestion
The NIST team has no long-term plan to maintain this project outside the context of the initiatives it supports, [OSCAL](https://pages.nist.gov/OSCAL) and [Metaschema](https://pages.nist.gov/metaschema), while we reserve the right to change and adapt to community needs in alignment with NIST's mission. A project goal is to enable the stabilization and socialization of the Metaschema technology, as evidenced by the use and support of *other* Metaschema implementations.
```

> 

Reviewer: These links are weird. They should be anchored on "OSCAL" and "Metaschema" instead of a URL.

```
git submodule update --init
```

Or do this in one step using `clone --recurse-submodules` as noted.

### Updating submodules

Submodule contents will be periodically updated. To ensure you have the latest commits for a configured submodule:

```
git submodule update --init --recursive
```

## Contact information

Principal Investigator: Wendell Piez, NIST (US National Institute of Standards and Technology) ITL/CSD (Information Technology Laboratory, Computer Security Division).

Email w e n d e l l (dot) p i e z (at) n i s t (dot) g o v.

As this initiative spins off from and supports the ITL/CSD [Metaschema](https://pages.nist.gov/metaschema) Project, originally and currently being developed in support of OSCAL, the [Open Security Controls Assessment Language](https://pages.nist.gov/OSCAL), feedback on this project will be gratefully received in those forums.

For OSCAL-related discussions see the [OSCAL Site](https://pages.nist.gov/OSCAL/contribute).

## Cite this work

Piez, Wendell (2023), Metaschema-XSLT. US National Institute of Standards and Technology (NIST). https://github.com/usnistgov/metaschema-xslt.

## Related projects

> See the [Metaschema Repository](https://github.com/usnistgov/metaschema) and its [Pages Site](https://pages.nist.gov/metaschema/) for a description and specification of the technology this software is designed to support.

Reviewer: ```suggestion
See the Metaschema [repository](https://github.com/usnistgov/metaschema) and [website](https://pages.nist.gov/metaschema/) for a description and specification of the technology this software is designed to support.
```

See the Project [Wiki](https://github.com/usnistgov/metaschema-xslt/wiki)  for documentation maintained on this site.

Projects in support of NIST Metaschema (at time of writing)

- https://github.com/usnistgov/metaschema-java
- https://github.com/usnistgov/metaschema-node

> - https://github.com/usnistgov/nmetaschema

Reviewer: None of these relate to *this* project, as they offer different implementations. They relate to metaschema in general. Maybe remove these here and add this in the metaschema repo?

Author: ```Suggestion
See the Project [Wiki](https://github.com/usnistgov/metaschema-xslt/wiki)  for documentation maintained on this site.
```

Projects currently known to be using NIST Metaschema

- [OSCAL: the Open Security Controls Assessment Language](https://pages.nist.gov/OSCAL/)

> 

Reviewer: Again, this project should not be speaking to use of NIST Metaschema, but it could suggest that this project is used by OSCAL.

> 

Author: ```Suggestion
See the Project [Wiki](https://github.com/usnistgov/metaschema-xslt/wiki)  for documentation maintained on this site.

Projects in support of NIST Metaschema (at time of writing)

- https://github.com/usnistgov/metaschema-java
- https://github.com/usnistgov/metaschema-node
- https://github.com/usnistgov/nmetaschema
```

## Required outline

This page includes all the following, as described by guidelines at https://raw.githubusercontent.com/usnistgov/opensource-repo/main/README.md

1. Software or Data description
   - Statements of purpose and maturity
   - Description of the repository contents
   - Technical installation instructions, including operating
     system or software dependencies
1. Contact information
   - PI name, NIST OU, Division, and Group names
   - Contact email address at NIST
   - Details of mailing lists, chatrooms, and discussion forums,
     where applicable
1. Related Material
   - URL for associated project on the NIST website or other Department
     of Commerce page, if available
   - References to user guides if stored outside of GitHub
1. Directions on appropriate citation with example text
1. References to any included non-public domain software modules,
   and additional license language if needed, *e.g.* [BSD][li-bsd],

>    [GPL][li-gpl], or [MIT][li-mit]

Reviewer: I think these are not valid links.

Author: okay will fix

# metaschema-xslt

> ## punchlist

Reviewer 2: Capitalize the title...

- [x] `CODEMETA`
- [x] `CONTRIBUTING.md`
- [x] `CODEOWNERS`
- [x] `LICENSE.md`
- [x] `CODE_OF_CONDUCT.md`
- [x] this file (see outline/requirements at end)

## License

See the [project license](LICENSE.md) in this repository.

This project is placed into the worldwide public domain.

## Software description

> Metaschema-XSLT is an XSLT-based implementation of the NIST (ITL/CSD) [Metaschema data modeling framework](https://github.com/usnistgov/metaschema).

Reviewer 3: Minor: is it not MIMF, per the official doc site updates?

```suggestion
Metaschema-XSLT is an XSLT-based implementation of the NIST (ITL/CSD) [Metaschema Information Modeling Framework](https://github.com/usnistgov/metaschema).
```

> Metaschema-XSLT is an XSLT-based implementation of the NIST (ITL/CSD) [Metaschema Information Modeling Framework](https://github.com/usnistgov/metaschema).

Reviewer 2: Metaschema Information Modeling Framework name does not appear in the metaschema repo. Should it be added in the Readme.md of metaschema for consistency?

Also, from this point down, the entire section appears to focus on MIMF and not Metaschema-XSLT.

Reviewer 3: Isn't it in the specification?

https://pages.nist.gov/metaschema/specification/#metaschema-specification

Author: Yup, and all over that Pages site. Alas, maintaining that site's readme is out of scope for this Issue.

Stress is therefore on:

- Transparency, testability and maintainability of processing on an open-source stack
- Conformance in support of Metaschema-based data interchange / interop
- Generality of support of Metaschema features

What we do not emphasize:

- Performance
- Portability beyond the XML/XSLT dependency stack (instead relying on other Metaschema initiatives)
- Use cases in isolation

### Origins

Formerly housed in the Metaschema repository, this code base traces the history of development of the Metaschema concept in the context of the OSCAL project. It was originally conceived as a demonstration and proof of concept, providing a bridge enabling JSON- and XML-based development in parallel over common problem sets and common data. Success in this effort led to a determination that multiple implementations of a platform-independent specification were needed, at which point this implementation was carved out into its own repository.

### Project sunset

> Currently there is no plan to continue maintaining this project beyond the retirement of its lead researcher. A project goal is to enable the stabilization and socialization of the Metaschema technology, as evidenced by the use and support of *other* Metaschema implementations.

Reviewer 2: The statement "Currently there is no plan to continue maintaining this project beyond the retirement of its lead researcher" brings forward a privacy concern since it points to the lead's age and plan to retire in the near future. 
I suggest:
NIST team has no long-term plan to maintain this project but reserves the right to change and adapt to community needs if aligned with NIST mission.

Author: Actually it doesn't say I plan to retire in the near future (and I don't), but still.

The best way to ensure long-term access to the code base is to clone or fork the git repository and archive a copy for study or use.

## Repository contents

`src` includes XSLT source code, with supporting infrastructure including ad-hoc testing

`support` includes dependent submodules with other static resources for configuration

> ## Installation and use

Reviewer 2: I suggest removing "and use" from the title since the next section is focusing on Using or Running it.

Author: "To run" is a subheading under "Installation and Use". But I'm trying for clarifying adjustments anyhow please check back after next commit. Will be much improved!

In particular, a convenient way to manage dependencies is to use Maven to manage a Java-based stack deploying SaxonHE and/or XML Calabash, as open-source XML processors. Scripts are provided for running under Maven.

### To run

The following generalized services are provided by the tools in this repository, separately or in combination

- XSD and JSON schema generation - `schema-gen` folder
- Converter XSLTs for metaschema-supported data - `converter-gen`
- Metaschema documentation `document`.

Scripts and stylesheets will be documented *in situ* using readmes and in line. Most scripts depend on Apache Maven supporting a Java runtime. Since XSLTs can call, import, include or read XSLTs from elsewhere in the repo, and sometimes do, keep the modules together: each folder on its own is *not* self-contained.

For testing, all XSpec scenarios (`*.xspec`) can be run in place to generate local test reports.

Users are also expected to call resources in this repository from their own scripts. Do this either by cloning, copying and modifying scripts here; by writing your own; or by adapting code into the XML/XSLT processing framework or stack of your choice.

> A convention is used indicating that an XProc (`*.xpl` file) or XSLT (`*.xsl`) intended to be invoked directly (that is, not only to be used as a module or component) is given a name entirely or partly in `ALL-CAPITALS`. For example, `src/schema-gen/METASCHEMA-ALL-SCHEMAS.xpl` is such an XProc pipeline (a step definition intended to be used directly).

Reviewer 3: Blocking: I would expect to find information about using `make ...` commands to show primary usage or how to properly use shell to invoke the stylesheets and pipeline files as canonical examples of these, no?

Author: Is this really blocking? This branch is intended to be a WIP placeholder for further work.

Reviewer 3: OK, for context this did come up on the different (but related) OSCAL release repo that if we expect someone other than Nikita and you to work on this (that was an expectation until we reshuffled priorities, Dmitry and I had opted to take it on but we moved away from that after 15 min or so), that stuff becomes important for making sure fresh eyes can get to work.

I will follow up in subsequent issues/dev work on the developer docs.

### Dependencies

As noted, the Saxon XSLT engine can be regarded as a *de facto* dependency - while this XSLT-conformant code should in principle run in any processor implementing the language. SaxonHE can be bundled using Maven or another Java packaging technology.

The [POM file](support/pom.xml) for Java/Maven configuration indicates the current tested version of Saxon. At time of writing, Saxon versions 10 and 11 are known to work with this codebase. When reporting bugs please include the version of your processor.

Developers interested in demonstrating the viability of these processes in different processors and environments are eagerly invited to participate in development of this tool or [related tools](#Related_projects).

### Git Client Setup

> See more on git setup on the page on [Contributing](CONTRIBUTING).

Reviewer 2: Suggest: See more on git setup on the [Contributing](CONTRIBUTING) page.

Clone the project:

```
git clone --recurse-submodules https://github.com/usnistgov/metaschema-xslt.git
```

Or to clone with submodule initialization:

```
git clone --recurse-submodules https://github.com/usnistgov/metaschema-xslt.git
```

### Initializing Git submodules

This GitHub repository makes use of Git submodules to mount other repositories as subdirectories. When cloning this repo for the first time, you need to initialize these.

```
git submodule update --init
```

Or do this in one step using `clone --recurse-submodules` as noted.

### Updating submodules

Submodule contents will be periodically updated. To ensure you have the latest commits for a configured submodule:

```
git submodule update --init --recursive
```

## Contact information

Principal Investigator: Wendell Piez, NIST (US National Institute of Standards and Technology) ITL/CSD (Information Technology Laboratory, Computer Security Division).

Email w e n d e l l (dot) p i e z (at) n i s t (dot) g o v.

> This initiative spins off from and supports the ITL/CSD [Metaschema](https://pages.nist.gov/metaschema) Project, originally and currently being developed in support of OSCAL, the [Open Security Controls Assessment Language](https://pages.nist.gov/OSCAL).

Reviewer 2: An identical or very similar statement already exists above. Suggest remove it from here or re-write it to say that any issue or question related to the MIMF or OSCAL should be posted on the respective GitHub repos.

For OSCAL-related discussions see the [OSCAL Site](https://pages.nist.gov/OSCAL/contribute).

> ## Cite this work

Reviewer 2: If you approve the site in MIDAS, I am suggesting adding the doi here.

Author: Definitely! I had not decided on proceeding just yet with the MIDAS listing but you are right, we should do so, and the DOI is easy enough to acquire.

In chicken-and-egg fashion it will be easier to get review approval in/for MIDAS after this merge, but I did make an Issue: https://github.com/usnistgov/metaschema-xslt/issues/4.
